data/mods/ssb/abilities.ts

```diff
@@ -42,10 +41,11 @@ export function changeSet(context: Battle, pokemon: Pokemon, newSet: SSBSet, cha
 		2 * pokemon.species.baseStats.hp + pokemon.set.ivs.hp + Math.floor(pokemon.set.evs.hp / 4) + 100
 	) * pokemon.level / 100 + 10);
 	const newMaxHP = pokemon.baseMaxhp;
-	pokemon.hp = newMaxHP - (pokemon.maxhp - pokemon.hp);
+	const percent = (pokemon.hp / pokemon.maxhp);
+	pokemon.hp = Math.round(newMaxHP * percent);
 	pokemon.maxhp = newMaxHP;
 	let item = newSet.item;
-	if (typeof item !== 'string') item = item[Math.floor(Math.random() * item.length)];
+	if (typeof item !== 'string') item = item[pokemon.battle.random(item.length)];
```

Comment:
```suggestion
	if (typeof item !== 'string') item = item[context.random(item.length)];
```
Context is battle, this is more clear.